tests/group_functions/group_joint_reduce.cpp.in

```diff
@@ -20,32 +20,31 @@
 
 #include "group_reduce.h"
 
+// clang-format off
+#cmakedefine CTS_TYPE @CTS_TYPE@
+#cmakedefine CTS_TYPE_NAME std::string("@CTS_TYPE_NAME@")
+// clang-format on
 using ReduceTypes = Types;
 
-// 2-dim Cartesian product of type lists
-using prod2 = product<std::tuple, ReduceTypes, ReduceTypes>::type;
-
 // hipSYCL has no implementation over sub-groups
-TEMPLATE_LIST_TEST_CASE("Group and sub-group joint reduce functions",
-                        "[group_func][type_list][dim]", ReduceTypes) {
+TEST_CASE(CTS_TYPE_NAME + " group and sub-group joint reduce functions",
+          "[group_func][type_list][dim]") {
   auto queue = once_per_unit::get_queue();
-  // Get binary operators from TestType
-  const auto Operators = get_op_types<TestType>();
-  const auto Type = unnamed_type_pack<TestType>();
-  for_all_combinations<invoke_joint_reduce_group>(Dims, Type, Operators, queue);
+  const auto Operators = get_op_types<CTS_TYPE>();
+  const auto RetType = unnamed_type_pack<CTS_TYPE>();
+  for_all_combinations<invoke_joint_reduce_group>(Dims, RetType, Operators,
+                                                  queue);
 }
 
 // hipSYCL has problems with 16-bit types for Ptr
-TEMPLATE_LIST_TEST_CASE("Group and sub-group joint reduce functions with init",
-                        "[group_func][type_list][dim]", prod2) {
+TEMPLATE_LIST_TEST_CASE(
+    CTS_TYPE_NAME + " group and sub-group joint reduce functions with init",
+    "[group_func][type_list][dim]", ReduceTypes) {
   auto queue = once_per_unit::get_queue();
-  using T = std::tuple_element_t<0, TestType>;
-  using U = std::tuple_element_t<1, TestType>;
 
-  // Get binary operators from T
-  const auto Operators = get_op_types<T>();
-  const auto RetType = unnamed_type_pack<T>();
-  const auto ReducedType = unnamed_type_pack<U>();
+  const auto Operators = get_op_types<CTS_TYPE>();
+  const auto RetType = unnamed_type_pack<CTS_TYPE>();
+  const auto ReducedType = unnamed_type_pack<TestType>();
```

Comment:
Am I right in assuming these test generator files are going to be used for `sycl::half` and `double` as well? If so, we should have a check in this and the other generator files that does an aspect check (see deleted files) if any of those types are being used.

Reply:
> I move aspect check to header files, because only these 2 types need aspect checking.

To be honest, I don't see connection between the types and the place where we check their support. Adding aspect check here seems more efficient. Checking in call operator happens later in the execution pipeline.

Having said that, I'm not sure if the overhead is significant, but I think all other tests check do aspect checking very early.

Reply:
Fixed in 3cc9e85